uaclient/clouds/aws.py

```diff
@@ -18,34 +21,60 @@
 class UAAutoAttachAWSInstance(AutoAttachCloudInstance):
 
     _api_token = None
+    _ip_addresss_type = None
+
+    def _get_imds_url_response(self):
+        headers = self._request_imds_v2_token_headers()
+        return util.readurl(
+            IMDS_URL.format(IMDS_IP_ADDRESS[self._ip_addresss_type]),
+            headers=headers,
+            timeout=1,
+        )
 
     # mypy does not handle @property around inner decorators
     # https://github.com/python/mypy/issues/1362
     @property  # type: ignore
     @util.retry(HTTPError, retry_sleeps=[1, 2, 5])
     def identity_doc(self) -> Dict[str, Any]:
-        headers = self._get_imds_v2_token_headers()
-        response, _headers = util.readurl(IMDS_URL, headers=headers)
+        response, _headers = self._get_imds_url_response()
         return {"pkcs7": response}
 
+    def _request_imds_v2_token_headers(self):
+        try:
+            headers = self._get_imds_v2_token_headers(ip_address_type="ipv4")
+        except HTTPError as e:
+            raise e
+        except Exception:
+            msg = (
+                "Failed to reach AWS IPv4 IMDS route.\n"
+                "Trying AWS IPv6 route instead ..."
+            )
+            logging.info(msg)
+            headers = self._get_imds_v2_token_headers(ip_address_type="ipv6")
+
+        return headers
+
     @util.retry(HTTPError, retry_sleeps=[1, 2, 5])
-    def _get_imds_v2_token_headers(self):
+    def _get_imds_v2_token_headers(self, ip_address_type="ipv4"):
```

Comment:
we probably don't need ip_address_type here, but ip_address itself because all we are doing is looking that value up anyway. And make it mandatory positional.
```suggestion
    def _get_imds_v2_token_headers(self, ip_address):
```